test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NetCorePackageReferenceProjectTests.cs

```diff
@@ -447,6 +456,1817 @@ public async Task GetInstalledVersion_WithAssetsFile_ChangingPackageSpec_Returns
+                {
+                    // Install packageB since packageA is already there.
+                    await nuGetPackageManager.InstallPackageAsync(netCorePackageReferenceProjects[i], packageB, new ResolutionContext(), new TestNuGetProjectContext(),
+                            sourceRepositoryProvider.GetRepositories(), sourceRepositoryProvider.GetRepositories(), CancellationToken.None);
+
+                    // This code added because nuGetPackageManager.InstallPackageAsync doesn't do updating ProjectSystemCache
+                    var installed = new LibraryDependency
+                    {
+                        LibraryRange = new LibraryRange(
+                            name: packageB.Id,
+                            versionRange: new VersionRange(packageB.Version),
+                            typeConstraint: LibraryDependencyTarget.Package),
+                    };
+
+                    var packageSpec = packageSpecs[i];
+                    packageSpec.TargetFrameworks.First().Dependencies.Add(installed);
+                    var projectRestoreInfo = ProjectJsonTestHelpers.GetDGSpecFromPackageSpecs(packageSpec);
+                    var projectNames = GetTestProjectNames(projectFullPaths[i], $"project{i}");
+                    projectCache.AddProjectRestoreInfo(projectNames, projectRestoreInfo, new List<IAssetsLogMessage>());
+                }
+
+                var initialInstalledPackages = await netCorePackageReferenceProjects[numberOfProjects - 1].GetInstalledPackagesAsync(CancellationToken.None);
+
+                var targetProjects = new List<NetCorePackageReferenceProject>()
+                {
+                    netCorePackageReferenceProjects[numberOfProjects-2] // Middle parent project.
+                };
+
+                // Act
+                var results = await nuGetPackageManager.PreviewProjectsInstallPackageAsync(
+                    targetProjects,
+                    packageB_UpgradeVersion,
+                    resolutionContext,
+                    testNuGetProjectContext,
+                    sourceRepositoryProvider.GetRepositories(),
+                    CancellationToken.None);
+                var actions = results.Select(a => a.Action).ToArray();
+                await nuGetPackageManager.ExecuteNuGetProjectActionsAsync(
+                    netCorePackageReferenceProjects,
+                    actions,
+                    testNuGetProjectContext,
+                    new SourceCacheContext(),
+                    CancellationToken.None);
+
+                // Assert
+                Assert.Equal(initialInstalledPackages.Count(), 2);
+                var builtIntegratedActions = actions.OfType<BuildIntegratedProjectAction>().ToList();
+                Assert.Equal(actions.Count(), targetProjects.Count());
+                Assert.Equal(actions.Count(), builtIntegratedActions.Count());
+                // Upgrade succeed for this middle parent project(with parent and childs).
```

Comment:
I see so many references to `.Count()` extension method which is O(n) time complexity. Use[ `.Count`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.count?view=netcore-3.1) instead.

Reply:
On some .NET runtimes, LINQ's `.Count()` tries to typecast to `ICollection` and then call `.Count`, so it's not O(n). But, just like this other comment thread: https://github.com/NuGet/NuGet.Client/pull/3559#discussion_r481375529

Basically, we should all be trying to avoid LINQ where possible. Use it when it increases readability, but be aware that it always has a performance cost. Even using `foreach` has a perf cost compared to a `for` loop, but the perf difference I think is smaller than avoiding LINQ.

Reply:
I think calling just once for `initialInstalledPackages.Count()` is ok, because we're evaluating only 1 time. If we convert to List/Array then by the time we call `.ToList()/ToArray()` it'll cost too. 
But for repeated use of .Count() or .Any() on same IEnumerable item will of course adds up.